onnxruntime/core/mickey/blk_q4/prepack_sm80.h

```diff
@@ -0,0 +1,317 @@
+// Licensed under the MIT License.
+
+#pragma once
+
+// #include <cuda.h>
+// #include <vector_types.h>
+// #include "cutlass/cutlass.h"
+// #include "cutlass/matrix_shape.h"
+// #include "cutlass/util/host_tensor.h"
+// #include "cutlass/util/reference/host/tensor_compare.h"
+// #include "cutlass/util/reference/host/tensor_copy.h"
+// #include "cutlass/util/reference/host/tensor_fill.h"
+
+#include "core/common/common.h"
+#include "core/util/matrix_layout.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+/**
+ * @brief Blockwise quantization methods
+ * @tparam ElementT       source data type, fp16
+ * @tparam block_size     number of elemenets quantized together
+ * @tparam qbits          number of bits in each quantized element
+ * @tparam Columnwise     true:  elements in a block come from one single column
+ *                        false: elements in a block come from one single row
+ */
+template <
+    typename ElementT,
+    int block_size,
+    int qbits,
+    bool Columnwise,
+    bool ExtraBoundsCheck = false>
+struct BlockwiseQuantization {
+  static_assert(qbits == 4, "Only 4b block quantization is supported!");
+  static_assert(sizeof(ElementT) == 2, "Only 16b floating point types are supported!");
+
+  using QuantBlocking =
+      std::conditional_t<Columnwise,
+                         MatrixShape<block_size, 1>,
+                         MatrixShape<1, block_size>>;
+
+  using ElementW = uint8_t;  // <- Weight is int4, uint8 for two of them
+  // We pack 4 weights into one 16b element, so we can leverage cutlass tile iterators
+  // for async share memory loading, and minimizing bank conflict during matrix loading
+  using ElementWPack = ElementT;
+  using LayoutWPack = ColumnMajorLayout;  // <- layout of packed weight, must be column major
+
+  // Current Ampere kernel use 8b zero point, need to shrink it to 4b in the future
+  using ElementQOffset = uint8_t;
+
+  // Layout of the quantization parameters (scales and zero points)
+  // Major on the dimension that has the most parameters per squarish weight block.
+  // E.g. for column-wise quantization, a [64, 64] block has [2, 64] parameters,
+  // where each row has more data, so we use row major layout so that warp threads
+  // can use less load instructions to load more parameters.
+  using LayoutQmeta =
+      typename std::conditional<Columnwise,
+                                RowMajorLayout, ColumnMajorLayout>::type;
+
+  /**
+   * @brief  Get quantized weight tensor dimensions.
+   * Actual weight type is int4, we use ElementW = uint8 to avoid possible compilation
+   * troubles. Since the layout is column major, we are packing 2 weights in a column
+   * into one int8
+   */
+  static inline auto get_quant_weights_shape(int rows, int columns) {
+    return make_Position((rows + 1) / 2, columns);
+  }
+
+  static inline auto get_quant_meta_shape(int rows, int columns) {
+    return make_Position(rows / QuantBlocking::kRow, columns / QuantBlocking::kColumn);
```

Comment:
>rows / QuantBlocking::kRow, columns / QuantBlocking::kColumn [](http://example.com/codeflow?start=25&length=60)

do you need to use ceiling here?

Reply:
due to limitation in one of the cutlass global memory iterators, this sm80 kernel can not handle 'half blocks'. Added verification at the beginning of prepack_weights()